### PR TITLE
Documentation on Memory and Storage profiling of IoT edge

### PR DESCRIPTION
**Pull request by ggjjj**
* Setup of the profiling platform
* Metrics used for profiling
* Summary of the numbers after profiling
* How compatibility scripts uses these metrics for doing memory and storage checks

### CODE REVIEW

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+IoT Edge Compatibility script performs many checks to figure out whether a platform has capabilities to run IoT Edge or not. This document is focused on how the profiling of IoT Edge is performed and how the profiling information is used to perform the memory and storage checks in the Compatibility script for platform validation.
+
+## Setup
+
+IoT Edge binaries like `aziot-edged`, `aziot-idenitityd` are installed in three different platforms like x86_64, aarch64  and armv7l with Ubuntu 20.04.
+A simple workload of three containers like edgeAgent, edgeHub and SimulatedTemperatureSensor were deployed. The profiling has been done by running the workload for a duration of **one hour**.
+Note: The metrics would drastically vary if the workload would have more container workloads other than the one specified in the setup. The version of the binaries and containers of IoT Edge are mentioned in the table below.
+
+|  Name | Version |
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md:11]: We should specify the exact versions we did our profiling on

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Memory and Storage Profiling of IoT Edge
+
+IoT Edge Compatibility script performs many checks to figure out whether a platform has capabilities to run IoT Edge or not. This document is focused on how the profiling of IoT Edge is performed and how the profiling information is used to perform the memory and storage checks in the Compatibility script for platform validation.
+
+## Setup
+
+IoT Edge binaries like `aziot-edged`, `aziot-idenitityd` are installed in three different platforms like x86_64, aarch64  and armv7l with Ubuntu 20.04.
+A simple workload of three containers like edgeAgent, edgeHub and SimulatedTemperatureSensor were deployed. The profiling has been done by running the workload for a duration of **one hour**.
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md:8]: ```A simple workload of Edge RunTime Containers (edgeAgent and edgeHub) along with SimulatedTemperatureSensor Module were deployed```

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+|                         |x86_64      |armv7l      |aarch64     |              
+|:-----------------------:|:----------:|:----------:|:----------:|
+|On Memory Binary Size    |54.24       |35.51       |26.62       |
+|On Memory Container Size |175.00      |164.53      |210.00      |
+|On Disk Binary Size      |42.39       |36.68       |36.68       |
+|On Disk Container Size   |254.96      |322.98      |322.60      |
+
+## Analysis
+
+The four metrics are used by the IoT Edge compatibility script to evaluate whethe the platform has enough memory and storage to run IoT edge. A buffer size of **50 MB** is observed with the current metrics. The compatibility script queries the `/proc/meminfo` file in the system to evaluate the memory checks and `df` command to evaluate the storage checks.
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md:46]: We dont need to mention details about buffer size. That's for internal use.

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+### Memory check
+* On Memory Binary Size
+	* The average dynamic memory consumed by the IoT edge binaries like aziot-edged, aziot-identityd, aziot-certd etc and it is obtained by using `ps` command.
+* On Memory Container Size
+	* The average dynamic memory consumed by the IoT edge workload containers like edgeAgent, edgeHub and SimulatedTemperatureSensor and it is obtained by `docker stats` command.
+
+### Storage check
+* On Disk Binary Size
+	* The average storage space consumed by the IoT edge binaries like aziot-edged, aziot-identityd, aziot-certd etc and it is obtained by calculating the binaries size using `stat` command in the system.
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md:31]: Storage space is not averaged, its static and we take a summation of the storage space required by the binaries.

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+* On Disk Container Size
+	* The average storage space consumed by the IoT edge workload containers like edgeAgent, edgeHub and SimulatedTemperatureSensor and it is obtained by `docker inspect` command. Only size of the unique layers of the containers are considered for this metric.
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md:33]: Same goes here

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# Memory and Storage Profiling of IoT Edge
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md:1]: We can simply call it Resource Profile for IoT Edge

File: platform-validation/docs/IoTEdge-Profiling.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+Note: The metrics would drastically vary if the workload would have more container workloads other than the one specified in the setup. The version of the binaries and containers of IoT Edge are mentioned in the table below.
+
+  
+
+| Name | Version |
+|:-----:|:-------:|
+| IoT Edge Service| 1.2.8|
+| IoT Identity service | 1.2.6 |
+| EdgeAgent | 1.2 |
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Profiling.md:20]: Can you specify the exact version of Edge Agent and Edge Hub as well?

File: platform-validation/docs/IoTEdge-Profiling.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+| aarch64 | TBD | 2 | 8 | 75 
+| x86_64 |Intel Xeon E5-2673 v4 @ 2.29 GHz 1 Processor, 1 Core, 2 Threads|  2 | 8 | 16 |
+
+## Metrics
+
+Two major metrics are observed from this  i.e memory and storage. In both the categories the binary and container size is profiled.
+
+### Memory check
+
+* Memory Usage (Resisent Set Size) of Binaries
```
Comment (nimanch) [inline platform-validation/docs/IoTEdge-Profiling.md:38]: *Resident Set Size*

File: platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+### Storage check
+* On Disk Binary Size
+	* The average storage space consumed by the IoT edge binaries like aziot-edged, aziot-identityd, aziot-certd etc and it is obtained by calculating the binaries size using `stat` command in the system.
+* On Disk Container Size
+	* The average storage space consumed by the IoT edge workload containers like edgeAgent, edgeHub and SimulatedTemperatureSensor and it is obtained by `docker inspect` command. Only size of the unique layers of the containers are considered for this metric.
+
+A summary of the metrics by profiling specified version of IoT edge in three platforms are as follows. The sizes are in **MB**
+
+|                         |x86_64      |armv7l      |aarch64     |
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply platform-validation/docs/IoTEdge-Memory-Storage-Profiling.md]: whenever the numbers change we change the version.

**Review by nimanch — APPROVED** [review]

**Review by nimanch — APPROVED** [review]

**Pull request merged**